File: app/src/main/java/com/babestudios/companieshouse/ui/filinghistory/FilingHistoryAdapter.java
package com.babestudios.companieshouse.ui.filinghistory;


import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.babestudios.companieshouse.R;
import com.babestudios.companieshouse.data.DataManager;
import com.babestudios.companieshouse.data.model.filinghistory.FilingHistoryList;

import butterknife.Bind;
import butterknife.ButterKnife;

public class FilingHistoryAdapter extends RecyclerView.Adapter<FilingHistoryAdapter.FilingHistoryViewHolder> {

	private FilingHistoryRecyclerViewClickListener mItemListener;

	private FilingHistoryList filingHistoryList = new FilingHistoryList();

	DataManager dataManager;

	FilingHistoryAdapter(Context c, FilingHistoryList filingHistoryList, DataManager dataManager) {
		mItemListener = (FilingHistoryRecyclerViewClickListener) c;
		this.filingHistoryList = filingHistoryList;
		this.dataManager = dataManager;
	}

	@Override
	public FilingHistoryViewHolder onCreateViewHolder(ViewGroup parent, int i) {
		View itemLayoutView = LayoutInflater.from(parent.getContext())
				.inflate(R.layout.filing_history_list_item, parent, false);

		return new FilingHistoryViewHolder(itemLayoutView);
	}

	@Override
	public void onBindViewHolder(FilingHistoryViewHolder viewHolder, int position) {

		viewHolder.lblDescription.setText(dataManager.filingHistoryLookup(filingHistoryList.items.get(position).description));
		viewHolder.lblDate.setText(filingHistoryList.items.get(position).date);
		viewHolder.lblCategory.setText(filingHistoryList.items.get(position).category);
		viewHolder.lblType.setText(filingHistoryList.items.get(position).type);

	}

	public long getItemId(int position) {
		return position;
	}

	@Override
	public int getItemCount() {
		return filingHistoryList.items.size();
	}

	class FilingHistoryViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
		@Bind(R.id.lblDate)
		TextView lblDate;
		@Bind(R.id.lblDescription)
		TextView lblDescription;
		@Bind(R.id.lblCategory)
		TextView lblCategory;
		@Bind(R.id.lblType)
		TextView lblType;

		FilingHistoryViewHolder(View itemView) {
			super(itemView);
			ButterKnife.bind(this, itemView);
			itemView.setOnClickListener(this);
		}

		@Override
		public void onClick(View v) {
			//mItemListener.searchResultItemClicked(v, this.getLayoutPosition(), filingHistoryList.items.get(getLayoutPosition()).title, filingHistoryList.items.get(getLayoutPosition()).companyNumber);
		}
	}

	interface FilingHistoryRecyclerViewClickListener {
		void searchResultItemClicked(View v, int position, String companyName, String companyNumber);
	}

	void addItems(FilingHistoryList filingHistoryList) {
		this.filingHistoryList.items.addAll(filingHistoryList.items);
		notifyDataSetChanged();
	}
}

File: app/src/main/java/com/babestudios/companyinfouk/ui/filinghistory/list/FilingHistoryAdapter.kt
package com.babestudios.companyinfouk.ui.filinghistory.list

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.babestudios.base.mvp.list.BaseViewHolder
import com.babestudios.companyinfouk.data.model.filinghistory.FilingHistoryItem
import com.jakewharton.rxbinding2.view.RxView
import io.reactivex.Observable
import io.reactivex.subjects.PublishSubject

class FilingHistoryAdapter internal constructor(
		private var visitables: List<FilingHistoryVisitable>,
		private val filingHistoryTypeFactory: FilingHistoryTypeFactory
) : RecyclerView.Adapter<BaseViewHolder<FilingHistoryVisitable>>() {

	interface FilingHistoryTypeFactory {
		fun type(filingHistoryItem: FilingHistoryItem): Int
		fun holder(type: Int, view: View): BaseViewHolder<*>
	}

	override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BaseViewHolder<FilingHistoryVisitable> {
		val view = LayoutInflater.from(parent.context).inflate(viewType, parent, false)
		val v = filingHistoryTypeFactory.holder(viewType, view) as BaseViewHolder<FilingHistoryVisitable>
		RxView.clicks(view)
				.takeUntil(RxView.detaches(parent))
				.map { v }
				.subscribe(itemClickSubject)
		return v
	}

	override fun onBindViewHolder(viewHolder: BaseViewHolder<FilingHistoryVisitable>, position: Int) {
		viewHolder.bind(visitables[position])
	}

	override fun getItemViewType(position: Int): Int {
		return visitables[position].type(filingHistoryTypeFactory)
	}

	private val itemClickSubject = PublishSubject.create<BaseViewHolder<FilingHistoryVisitable>>()

	fun getViewClickedObservable(): Observable<BaseViewHolder<FilingHistoryVisitable>> {
		return itemClickSubject
	}

	override fun getItemId(position: Int): Long {
		return position.toLong()
	}

	override fun getItemCount(): Int {
		return visitables.size
	}

	fun updateItems(visitables: List<FilingHistoryVisitable>) {
		this.visitables = visitables
		notifyDataSetChanged()
	}
}
